StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs

```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+                    if (locations.Length > 0)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, locations.First()));
+                    }
+                }
+            }
+        }
+
+        private void HandleDelegateDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            DelegateDeclarationSyntax declaration = context.Node as DelegateDeclarationSyntax;
+
+            bool isDefinedInClassOrStruct = IsDefinedInClassOrStructRecursive(declaration.Parent);
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, isDefinedInClassOrStruct ? SyntaxKind.PrivateKeyword : SyntaxKind.PublicKeyword))
+            {
+                var leadingTrivia = declaration.GetLeadingTrivia();
+                var commentTrivia = leadingTrivia.FirstOrDefault(x => x.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia));
+                if (XmlCommentHelper.IsMissingOrEmpty(commentTrivia))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                }
+            }
+        }
+
+        private void HandleEventDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            EventDeclarationSyntax declaration = context.Node as EventDeclarationSyntax;
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
+            {
+                var leadingTrivia = declaration.GetLeadingTrivia();
+                var commentTrivia = leadingTrivia.FirstOrDefault(x => x.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia));
+                if (XmlCommentHelper.IsMissingOrEmpty(commentTrivia))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                }
+            }
+        }
+
+        private void HandleEventFieldDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            EventFieldDeclarationSyntax declaration = context.Node as EventFieldDeclarationSyntax;
+            var variableDeclaration = declaration?.Declaration;
+
+            if (variableDeclaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
+            {
+                var leadingTrivia = declaration.GetLeadingTrivia();
+                var commentTrivia = leadingTrivia.FirstOrDefault(x => x.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia));
+                if (XmlCommentHelper.IsMissingOrEmpty(commentTrivia))
+                {
+                    var locations = variableDeclaration.Variables.Select(v => v.Identifier.GetLocation()).ToArray();
+                    if (locations.Length > 0)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, locations.First()));
```

Comment:
I believe the warning should be reported for each event, as opposed to just the first one.